lib/pbench/client/types.py

```diff
@@ -0,0 +1,68 @@
+from pathlib import Path
+import re
+from typing import Any, Union
+
+# A set of types defined to conform to the semantic definition of a JSON
+# structure with Python syntax.
+JSONSTRING = str
+JSONNUMBER = Union[int, float]
+JSONVALUE = Union["JSONOBJECT", "JSONARRAY", JSONSTRING, JSONNUMBER, bool, None]
+JSONARRAY = list[JSONVALUE]
+JSONOBJECT = dict[JSONSTRING, JSONVALUE]
+JSON = JSONVALUE
+
+
+class JSONMap:
+    """This constructs an object with attributes from a JSON object to make
+    access a bit more convenient. E.g., in Javascript we can access
+    object.attribute as well as object["attribute"] ... this mapper provides
+    the same convenience for Python.
+
+    Note that we don't try to set any attributes if the JSON dict contains any
+    key which isn't a Python symbol.
+    """
+
+    SYMBOL = re.compile(r"^[\w_]+$")
+
+    def __init__(self, json: JSONOBJECT):
+        """Save the JSON payload; if all JSON keys are legal Python symbols,
+        also enable direct access to them as attributes.
+
+        Args:
+            json:   JSON dictionary
+        """
+        self.json = json
+        if all(self.SYMBOL.match(k) for k in json.keys()):
+            for k, v in json.items():
+                setattr(self, k, JSONMap(v) if isinstance(v, dict) else v)
+
+    def __getitem__(self, key: str) -> Any:
+        """Regardless of possible attribute mapping, allow access to any JSON
+        key by index.
```

Comment:
Interesting, although I don't think it's worth requiring another package. It suggests that it might work with non-Python names, but having to use `hasattr` and `getattr` somewhat defeats the utility; whereas my minimal mapping here just stops at a `dict` with non-Python names.